pkg/controller/container-runtime-config/container_runtime_config_controller.go

```diff
@@ -616,10 +616,39 @@ func (ctrl *Controller) syncContainerRuntimeConfig(key string) error {
 		}
 		glog.Infof("Applied ContainerRuntimeConfig %v on MachineConfigPool %v", key, pool.Name)
 	}
+	if err := ctrl.cleanUpDuplicatedMC(); err != nil {
+		return err
+	}
 
 	return ctrl.syncStatusOnly(cfg, nil)
 }
 
+// cleanUpDuplicatedMC removes the MC of uncorrected version if format of its name is non-suffixed generate-xxx.
+// BZ 1955517: upgrade when there are more than one configs, these non-suffixed MC will be duplicated
+// by upgraded MC with suffixed name (func getManagedKeyCtrCfg()) and fails the upgrade.
+func (ctrl *Controller) cleanUpDuplicatedMC() error {
+	ctrcfgMCSuffix := "generated-containerruntime"
+	// Get all machine configs
+	mcList, err := ctrl.client.MachineconfigurationV1().MachineConfigs().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("error listing containerruntime machine configs: %v", err)
+	}
+	for _, mc := range mcList.Items {
+		mcCtrlVersion := mc.Annotations[ctrlcommon.GeneratedByControllerVersionAnnotationKey]
+		if !strings.HasSuffix(mc.Name, ctrcfgMCSuffix) {
+			continue
+		}
+		// delete the mc if its non-suffixed and degraded
```

Comment:
I feel like I'm missing some information, but here are a couple of notes: 
- above we're skipping if it doesn't have the suffix, but the comment here says we're deleting if it doesn't have the suffix suffixed. 
- also, the comment says it will delete if degraded, but I'm not seeing us check if it is degraded.

Reply:
I should make the commet more clear, we will delete the MC without number suffix, those MC usually has ctrcfgMCSuffix.
The version.Hash is check the degradation.

Reply:
@QiWang19 I don't think we should just restrict this to non-suffixed MCs. We should delete any MC that is still using the old ignition version regardless of if it has a suffix or not.
I would suggest renaming `ctrcfgMCSuffix` to something like `generatedCtrcfg`  and check if the MC has that in its name.

Reply:
fixed